controller/src/main/java/io/pravega/controller/server/rpc/grpc/v1/ControllerServiceImpl.java

```diff
@@ -131,147 +154,239 @@ public void getSegments(GetSegmentsRequest request, StreamObserver<SegmentsAtTim
+                            StreamObserver<CreateScopeStatus> responseObserver) {
         log.info("createScope called for scope {}.", request.getScope());
-        processResult(controllerService.createScope(request.getScope()),
-                      responseObserver);
+        authenticateExecuteAndProcessResults(v -> checkAuthorization(request.getScope(), PravegaAuthHandler.PravegaAccessControlEnum.READ_UPDATE),
+                () -> controllerService.createScope(request.getScope()),
+                responseObserver);
     }
 
     @Override
     public void deleteScope(ScopeInfo request,
-            StreamObserver<DeleteScopeStatus> responseObserver) {
+                            StreamObserver<DeleteScopeStatus> responseObserver) {
         log.info("deleteScope called for scope {}.", request.getScope());
-        processResult(controllerService.deleteScope(request.getScope()),
-                      responseObserver);
+        authenticateExecuteAndProcessResults(v -> checkAuthorization(request.getScope(), PravegaAuthHandler.PravegaAccessControlEnum.READ_UPDATE),
+                () -> controllerService.deleteScope(request.getScope()),
+                responseObserver);
+    }
+
+    @Override
+    public void getDelegationToken(io.pravega.controller.stream.api.grpc.v1.Controller.StreamInfo request,
+                                   io.grpc.stub.StreamObserver<io.pravega.controller.stream.api.grpc.v1.Controller.DelegationToken> responseObserver)  {
+        log.info("createStream called for stream {}/{}.", request.getScope(),
+                request.getStream());
+        authenticateExecuteAndProcessResults(v -> checkAuthorizationWithToken(request.getScope() + "/" +
+                        request.getStream(), PravegaAuthHandler.PravegaAccessControlEnum.READ_UPDATE),
+                () -> CompletableFuture.completedFuture(Controller.DelegationToken
+                        .newBuilder()
+                        .setDelegationToken(getCurrentDelegationToken())
+                        .build()),
+                responseObserver);
     }
 
     // Convert responses from CompletableFuture to gRPC's Observer pattern.
-    private static <T> void processResult(final CompletableFuture<T> result, final StreamObserver<T> streamObserver) {
-        result.whenComplete(
-                (value, ex) -> {
-                    log.debug("result = " + (value == null ? "null" : value.toString()));
+    private static <T> void authenticateExecuteAndProcessResults(Predicate<Void> authenticator,
+                                                                 Supplier<CompletableFuture<T>> call, final StreamObserver<T> streamObserver) {
+        if (authenticator.test(null)) {
+            CompletableFuture<T> result = call.get();
+            result.whenComplete(
+                    (value, ex) -> {
+                        log.debug("result = " + (value == null ? "null" : value.toString()));
```

Comment:
please use the `{}` string formatting. That way, if `debug` log is not enabled, we don't have to concatenate strings every time we run this code. 

Also, no need to call `toString` - same reason.

Reply:
Fixed.